dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java

```diff
@@ -239,46 +243,75 @@ public PerHostClientFactory perHost(String serviceName) {
+                        boolean isSelf = targetUri
+                                .resolvedAddress()
+                                .map(selfAddresses::contains)
+                                .orElse(false);
 
                         // subtle gotcha here is that every single one of these has the same channelName,
                         // which means metrics like the QueuedChannel counter will end up being the sum of all of them.
-                        list.add(cache.getNonReloadingChannel(
-                                params,
-                                singleUriServiceConf,
-                                ImmutableList.of(targetUri),
-                                channelName,
-                                OptionalInt.of(i)));
+                        map.put(
+                                new PerHostTarget(targetUri, isSelf),
+                                cache.getNonReloadingChannel(
+                                        params,
+                                        singleUriServiceConf,
+                                        ImmutableList.of(targetUri),
+                                        channelName,
+                                        OptionalInt.of(i)));
                     }
-                    return list.build();
+                    return map.buildKeepingLast();
                 });
 
         return new PerHostClientFactory() {
             @Override
             public Refreshable<List<Channel>> getPerHostChannels() {
-                return perHostDialogueChannels.map(ImmutableList::copyOf);
+                return perHostChannels.map(channels -> ImmutableList.copyOf(channels.values()));
+            }
+
+            @Override
+            public Refreshable<Map<PerHostTarget, Channel>> getNamedPerHostChannels() {
+                return perHostChannels;
             }
 
             @Override
             public <T> Refreshable<List<T>> getPerHost(Class<T> clientInterface) {
-                return getPerHostChannels().map(channels -> channels.stream()
-                        .map(chan -> Reflection.callStaticFactoryMethod(clientInterface, chan, params.runtime()))
+                return perHostChannels.map(channels -> channels.values().stream()
+                        .map(channel -> Reflection.callStaticFactoryMethod(clientInterface, channel, params.runtime()))
                         .collect(ImmutableList.toImmutableList()));
             }
 
+            @Override
+            public <T> Refreshable<Map<PerHostTarget, T>> getNamedPerHost(Class<T> clientInterface) {
+                return perHostChannels.map(channels -> channels.entrySet().stream()
+                        .collect(ImmutableMap.toImmutableMap(
+                                Map.Entry::getKey,
+                                entry -> Reflection.callStaticFactoryMethod(
+                                        clientInterface, entry.getValue(), params.runtime()))));
```

Comment:
slight preference to avoid `stream()` where possible
```java
ImmutableMap.copyOf(Maps.transformValues(
    channels,
    channel -> Reflection.callStaticFactoryMethod(clientInterface, entry.getValue(), params.runtime())));
```

Reply:
Why is this code better?

Reply:
Streams are difficult for hotspot to optimize, and I find to-map stream collectors a bit more difficult to read at a glance (this makes it more clear to devs that the keys aren't being modified, and the transformValues call makes it clear to the internals that the resulting ImmutableMap can be pre-sized).

Almost certainly won't make a measurable difference in practice, my goal in the comment is to share how I'm thinking about this, it's entirely up to you whether or not you'd like to change the snippet :-)